app/src/ingest_edd_web.py

```diff
@@ -86,27 +104,73 @@ def _create_chunks(
             logger.warning("Skipping duplicate URL: %s", item["url"])
             continue
 
-        name = item["title"]
-        logger.info("Processing: %s (%s)", name, item["url"])
+        logger.info("Processing: %s", item["url"])
         urls_processed.add(item["url"])
 
         content = item.get("main_content", item.get("main_primary"))
-        assert content, f"Item {name} has no main_content or main_primary"
+        assert content, f"Item {item['url']} has no main_content or main_primary"
 
-        document = Document(name=name, content=content, source=item["url"], **doc_attribs)
+        document = Document(name=item["title"], content=content, source=item["url"], **doc_attribs)
         chunks, splits = _chunk_page(document, content)
+        logger.info("Split into %d chunks: %s", len(chunks), item["title"])
         result.append((document, chunks, splits))
     return result
 
 
+USE_MARKDOWN_TREE = True
+
+
+class EddChunkingConfig(ChunkingConfig):
+    def __init__(self) -> None:
+        super().__init__(app_config.sentence_transformer.max_seq_length)
+
+    def text_length(self, text: str) -> int:
+        return len(tokenize(text))
+
+
 def _chunk_page(
     document: Document, content: str
 ) -> tuple[Sequence[Chunk], Sequence[SplitWithContextText]]:
     splits: list[SplitWithContextText] = []
-    for headings, text in split_markdown_by_heading(f"# {document.name}\n\n" + content):
-        # Start a new split for each heading
-        section_splits = _split_heading_section(headings, text)
-        splits.extend(section_splits)
+    if USE_MARKDOWN_TREE:
+        chunking_config = EddChunkingConfig()
+
+        # Fix markdown formatting that causes markdown parsing errors
+        # '. . .' is parsed as sublists on the same line
+        # in https://edd.ca.gov/en/uibdg/total_and_partial_unemployment_tpu_5/
+        content = content.replace(". . .", "...")
+        # '. * ' is parsed as sublists; incorrect markdown from scraping
+        # in https://edd.ca.gov/en/about_edd/your-benefit-payment-options/
+        content = content.replace(". *\n", ". \n")
+        # nested sublist '+' created without parent list; incorrect markdown from scraping?
+        # in https://edd.ca.gov/en/disability/Employer_Physician-Practitioner_Automated_Phone_Information_System/
+        content = content.replace("* + ", "    + ")
+
+        try:
+            tree = create_markdown_tree(content, doc_name=document.name, doc_source=document.source)
+            tree_chunks = chunk_tree(tree, chunking_config)
+
+            for chunk in tree_chunks:
+                split = SplitWithContextText(
+                    chunk.headings, chunk.markdown, chunk.context_str, chunk.embedding_str
+                )
+                assert split.token_count == chunk.length
```

Comment:
Am I following this right:
 - `chunk_tree` returns ProtoChunk
 - There is a 1:1 mapping of ProtoChunk -> SplitWithContextText
 - There is then a 1:1 mapping from SplitWithContextText -> Chunk

and from there:
 - SplitWithContextText is only used by the `if not USE_MARKDOWN_TREE` branch,
 - which we plan to deprecate in the future after adopting this chunking approach?
 
---

Then I think an approach to encapsulate the chunking logic here (ingestion/markdown_chunking.py and ingestion/markdown_tree.py) is:

1. Expose from ingestion a function `def chunk_markdown(markdown: str) -> list[Chunk]`
2. Update `ingest_edd_web.py` to only call `chunk_markdown`

That also lets us define integration-level tests that illustrate (and enforce) the behavior of the encapsulated code: we can write a set of pairs of markdown input strings, and the expected list of output Chunks, and compare the two sets.

I'm not suggesting we do that as part of this PR but I do think it'd be good to follow up: those integration-level tests would make it clearer for other engineers what the behavior of the logic is, and also make it easier to refactor the code (whether to update the behavior or otherwise.)

Reply:
> I do think it'd be good to follow up

(to clarify I'm not saying this needs to be you Yoom!)

Reply:
Yes, except:

> SplitWithContextText is only used by the if not USE_MARKDOWN_TREE branch,

SplitWithContextText is used in both `if` branches:
https://github.com/navapbc/labs-decision-support-tool/pull/121/files#diff-456c2fb3cff98badccca05b82109f1bab9c91a077cdfbc21245de5db8fa69b8aR154

SplitWithContextText is the commonality between the `if` branches. `ProtoChunk` is new and only used in the tree-based chunking code.

The encapsulation idea sounds good. There's definitely more refactoring that can be done outside the tree-chunking code!

Reply:
> SplitWithContextText is used in both if branches

What I mean is that it's not actually _doing_ anything in the new `if USE_MARKDOWN_TREE` branch. The other branch calls some SplitWithContextText functionality like `SplitWithContextText.add_if_within_limit`, etc. So after we switch to the new functionality we can drop this part of the code.

Reply:
Yes!